sycl/include/CL/sycl/kernel_handler.hpp

```diff
@@ -0,0 +1,72 @@
+//==------ kernel_handler.hpp -- SYCL standard header file -----*- C++ -*---==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace detail {
+
+// This guard is needed because the libsycl.so can compiled with C++ <=14
+// while the code requires C++17. This code is not supposed to be used by the
+// libsycl.so so it should not be a problem.
+#if __cplusplus > 201402L
+template <auto &S> struct specialization_id_name_generator {};
+#endif
+
+} // namespace detail
+
+/// Reading the value of a specialization constant
+///
+/// \ingroup sycl_api
+class kernel_handler {
+public:
+#if __cplusplus > 201402L
+  template <auto &S>
+  typename std::remove_reference_t<decltype(S)> get_specialization_constant() {
+#ifdef __SYCL_DEVICE_ONLY__
+    return getSpecializationConstantOnDevice<S>();
+#else
+    throw cl::sycl::feature_not_supported(
+        "kernel_handler::get_specialization_constant() is not supported by "
+        "host device.",
+        PI_INVALID_OPERATION);
+#endif // __SYCL_DEVICE_ONLY__
+  }
+#endif // __cplusplus > 201402L
+
+  void __init_specialization_constants_buffer(
+      char *_SpecializationConstantsBuffer = nullptr) {
+    SpecializationConstantsBuffer = _SpecializationConstantsBuffer;
+  }
+
+private:
+#ifdef __SYCL_DEVICE_ONLY__
+  template <auto &S, typename T = std::remove_reference_t<decltype(S)>,
+            std::enable_if_t<std::is_scalar_v<T>>>
+  T getSpecializationConstantOnDevice() {
+    const char *SymbolicID = __builtin_unique_stable_name(
+        detail::specialization_id_name_generator<S>);
+    return __sycl_getScalar2020SpecConstantValue<T>(
```

Comment:
Declaration is here: https://github.com/intel/llvm/pull/3353/files#diff-622b2db16138a17764c8376940c1bb4f5d748a395849f21f7d9635ab78752e6eR36